### DISCUSSION TITLE
Borg check question: for SSHFS repo, how much data is transferred for each option?

### DISCUSSION THREAD

**Discussion by AndrewDDavis** [borgbackup/borg#5782] [Q&A] [answered]
I'm backing up to an older NAS (MyCloud EX4) with glibc that is too old for the borg binaries, so I'm using SSHFS. Also there is a 100 Mbit router in the network, so data transfers are limited to about 10.5 MB/s. The repo size is ~ 1 TB, so transfer of all the data should take ~ 28 hours.

Recently I got integrity errors when backing up, so I'm running `borg check` commands. I was surprised that `borg check --repository-only` took more than a day to complete, which likely means _all_ of the repo data was transferred over the network. I was expecting that `--repository-only` would need somewhat less than the full data to do its checks. My question is: how much data will borg transfer over SSHFS to perform its checks, with `--repository-only`, `--archives-only`, or `--verify-data`?

**Comment by ThomasWaldmann** [comment, marked as answer]
The repository part of the check will read all chunks in the repo (all contents of the segment files) to verify the crc32 checksums of the stored chunks and also to build a known-good fresh repo index.

The archives part will then iterate over all archives and do misc. checks, e.g. whether the referenced chunks are present in the repo index - also it will build a new known-good chunks index (that has refcount and size infos for each chunk).

--verify-data will also do a cryptographic check, whether the chunk contents are authentic. this will read a lot of data.

More details are in the borg check docs.

About integrity errors:

- if the repo was ever used with an unpatched borg < 1.1.11 make sure to read and follow the advisory at top of the borg changelog. borg 1.1.9 in debian/ubuntu has been patched with the fix for that issue.

- check your hardware, hw issues are a common cause for integrity errors. RAM, disk, ...

**Reply by AndrewDDavis** [reply]
Thanks for that further explanation on the various check flavours. I think my confusion stemmed from the docs saying "The archive checks can be time consuming ...", which to me implied _more_ time consuming than the repository check. But in my experience, whether running over `ssh:` or using SSHFS, checking the repository takes much longer than checking an archive. And the docs say the `--verify-data` option will be very slow; but when running over SSHFS, _all_ of the repository data has already been transferred over the network for the repository part of the check. Will `--verify-data` actually take longer in this case?

Also thanks very much for the note about borg versions -- I had been using 1.1.9 on Debian stable, which would have been patched I suppose, but I installed 1.1.15 from backports anyway, and went through the cache deletion, now doing another full `check --repair`. The last check I ran with 1.1.9 identified a _missing file chunk_ from my most recent archive, so that's not good -- I had thought that the repo corruption was caused by an interrupted backup, but maybe there are deeper problems. I will run the system diagnostics on the EX4 when this borg check finishes.

**Reply by ThomasWaldmann** [reply]
Time for archive check grows linearly with the amount of archives (assuming they are all of similar content, amount of files, volume of data chunks). But if you do not use `--verify-data` it is mostly a metadata check.

There is no local caching, so the repo check already having read all the repo data does not help the archives check, it will just read and check again (in a different way). Maybe also showing linear growth with archive count.

If there is a chunk missing from your most recent archive, you have a good chance that your source data still reproduces this chunk if you do a backup. So do a check --repair, another backup, and another check --repair - maybe it will heal the missing chunk then.

Interrupted backups are unlikely to corrupt the repo due to its log-like structure (always append) and borg's transactional behaviour. Either stuff is completed / valid and committed. Or it is not (e.g. due to interruption) and will be just rolled back to the last valid (committed) state next time you use the repo.

**Reply by AndrewDDavis** [reply]
Thanks, that's helpful -- I think the part I was missing was that `--verify-data` would apply to the archive check, as well as the repo check. And I will try your advice on doing another `check --repair` after the next backup. The missing chunk was on a file that hasn't changed in years, so I'm hopeful it can be healed.

**Reply by ThomasWaldmann** [reply]
--verify-data does **not** change how the **repository part** of the check works. for the **archives part** of the check, it adds a cryptographic verification of the chunk contents (without that option, it only checks for the chunks presence).

**Reply by AndrewDDavis** [reply]
Thanks for all your help with this -- since the last, I have done a lot of hard-drive maintenance on both the source and NAS side using smartctl and other tools. Borg helped me realize that there were problems on the drives (on both ends!) that were not becoming apparent until _all_ the data was checked. I'm now getting clean reports from the full repository checks (previous problems were healed), and closely monitoring the health of these drives.

IMO, the man page for borg-check makes it clear that `--verify-data` applies only to the archives part under the option's description:

> --verify-data:  perform cryptographic archive data integrity verification (conflicts with --repository-only)

However, this could be made more clear in the main "description" section of the man page, and the online documentation for borg, by making the paragraph about `--verify-data` a bullet point of the archives check section, and including the part about "conflicts with --repository-only".

**Reply by ThomasWaldmann** [reply]
@AndrewDDavis agreed, filed there: https://github.com/borgbackup/borg/issues/5808
